### ISSUE TITLE
How to wait till page reloads/redirects after clicking button

### ISSUE THREAD

**Issue by infixint943** [go-rod/rod#184]
Currently, I do something like this:
```
page.Element(".submit").Click()
time.Sleep(time.Millisecond*300)
page.WaitLoad()
// do next stuff here
```

Is there any better way than this? One way I tried (but failed was)
```
wait := page.WaitRequestIdle()
page.Element(".submit").Click()
wait()
// do next steps ...
```

**Comment by ysmood** [comment]
Can you provide minimal code for us to reproduce your issue?

**Comment by ysmood** [comment]
If a page is like this, I want to click the button, how can you define what is page loaded?

```html
<html>
    <body>
        
    </body>

    <script>
        setTimeout(() => {
            document.body.innerHTML = '<button>click</button>'
        }, 3000)
    </script>
</html>
```

In this case, the race selectors is the only choice for me.

**Comment by infixint943** [comment]
Here is the entire code.

```
page := Browser.Page("https://codeforces.com/enter").WaitLoad()
page.Element("#handleOrEmail").Input(usr)
page.Element("#password").Input(passwd)
page.Element(".submit").Click()
time.Sleep(time.Millisecond*300)
page.WaitLoad()

// Check if in the home page now
page.Screenshot("")
```

**Comment by ysmood** [comment]
Don't forget to log our conversation in discord 😃

**Comment by infixint943** [comment]
Thanks for the answer @ysmood.
For those interested, the answer is running running a race selector `page.ElementMatches()` for some unique value in either of the 2 pages. In this case, it is `page.ElementMatches("a", "Logout", "span", "Invalid handle/email or password")` which runs a regex search on the select class.

**Comment by zplzpl** [comment]
I have the same question. Regarding how to get the elements of the new page after the form is submitted, the page redirect to a new page, I found that I can't get it.

**Comment by zplzpl** [comment]
i don't found page.ElementMatches().the lack of documentation confuses me. 

This is still a very common scenario.

form submit req -> resp redirect new page

**Comment by zplzpl** [comment]
```
	go page.EachEvent(func(e *proto.NetworkResponseReceived) {
		fmt.Println(json.Marshal(e)) // not any print
	}, func(e *proto.NetworkRequestWillBeSent) {
		fmt.Println(json.Marshal(e)) // not any print
	})

	url := "https://www.xxxx.com/login"
	page.MustNavigate(url).MustWaitLoad()

	_ = page.MustElement("#login").MustClick().Input("root")
	_ = page.MustElement("#password").MustClick().MustInput("root").MustPress(input.Enter)

	page.MustWaitLoad()

```

**Comment by infinitepr0** [comment]
You have 2 options:
1. explicitly wait for some particular element on login (page.MustElementR(`a`, `Welcome`) for example, which is present only once logged in).

2. Wait till page loads after clicking the button: `page.MustElement(`.submit`).MustClick().MustWaitInvisible().MustWaitLoad()`.

Option 1 is what you'd aim to do, if the webpage doesn't reload on submit. Plus it's faster, if you know exactly what you're searching for. Option 2 is suitable for all other cases.

**Comment by infinitepr0** [comment]
Also, the API changed a lot since when this issue was filed, you are looking for the method `page.HasR()`

**Comment by zplzpl** [comment]
Why can't capture redirect events and responses?

**Comment by ysmood** [comment]
@zplzpl If you are not comfortable using English please use https://www.deepl.com/ to translate your mother tongue.

> Why can't capture redirect events and responses?

Yes, Rod supports it, please use the search: https://github.com/go-rod/rod/search?q=redirect&type=issues

**Comment by zplzpl** [comment]
> @zplzpl If you are not comfortable using English please use https://www.deepl.com/ to translate your mother tongue. Your words don't make sense to me 😂
> 
> Please raise another issue for your question, yours is not related to the current one.

Thanks for the suggestion, I'll reopen an issue.

**Comment by tim0991** [comment]
if page redirect by js with settimeout  u can do this
```golang
for {
        // if u use MustElment to find a nonexistent elment will hang up even with out for
	if page.MustHas("#error_display") {
		fmt.Println("loginErr")
	}
	info := page.MustInfo()
	if strings.HasPrefix(info.URL, "other domain") {
		fmt.Println("loginSuccess")
		break
	}
	time.Sleep(time.Second)
}
```
